File: app/src/main/java/com/candy/capture/activity/VideoPlayerActivity.java
package com.candy.capture.activity;

import android.annotation.SuppressLint;
import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.Point;
import android.graphics.SurfaceTexture;
import android.graphics.drawable.ColorDrawable;
import android.media.MediaPlayer;
import android.os.Build;
import android.support.v7.app.ActionBar;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.util.Log;
import android.view.MotionEvent;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.candy.capture.R;
import com.candy.capture.model.MediaPlayState;
import com.candy.capture.util.DensityUtil;
import com.candy.capture.util.TimeUtil;
import com.candy.capture.util.TipsUtil;

import java.io.IOException;

public class VideoPlayerActivity extends BaseActivity implements View.OnClickListener {
    private static final String TAG = "VideoPlayerActivity";

    public static final String EXTRA_MEDIA_PATH_KEY = "extra_media_path";

    /**
     * Whether or not the system UI should be auto-hidden after
     * {@link #AUTO_HIDE_DELAY_MILLIS} milliseconds.
     */
    private boolean mAutoHide = true;

    /**
     * If {@link #mAutoHide} is set, the number of milliseconds to wait after
     * user interaction before hiding the system UI.
     */
    private static final int AUTO_HIDE_DELAY_MILLIS = 3000;
    /**
     * Some older devices needs a small delay between UI widget updates
     * and a change of the status and navigation bar.
     */
    private static final int UI_ANIMATION_DELAY = 300;

    private final Handler mHideHandler = new Handler();
    private final Runnable mHidePart2Runnable = new Runnable() {
        @SuppressLint("InlinedApi")
        @Override
        public void run() {
            // Delayed removal of status and navigation bar

            // Note that some of these constants are new as of API 16 (Jelly Bean)
            // and API 19 (KitKat). It is safe to use them, as they are inlined
            // at compile-time and do nothing on earlier devices.
            mSurfaceTv.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE
                    | View.SYSTEM_UI_FLAG_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);
        }
    };
    private final Runnable mShowPart2Runnable = new Runnable() {
        @Override
        public void run() {
            // Delayed display of UI elements
            ActionBar actionBar = getSupportActionBar();
            if (actionBar != null) {
                actionBar.show();
            }
            mProgressView.setVisibility(View.VISIBLE);
            mControllerView.setVisibility(View.VISIBLE);

            delayedHide(AUTO_HIDE_DELAY_MILLIS);
        }
    };
    private final Runnable mHideRunnable = new Runnable() {
        @Override
        public void run() {
            hide();
        }
    };
    private final Runnable mUpdateProgressRunnable = new Runnable() {
        @Override
        public void run() {
            if (mMediaPlayer != null) {
                mProgressSb.setProgress(mMediaPlayer.getCurrentPosition());
            }
        }
    };
    /**
     * Touch listener to use for in-layout UI controls to delay hiding the
     * system UI. This is to prevent the jarring behavior of controls going away
     * while interacting with activity UI.
     */
    private final View.OnTouchListener mDelayHideTouchListener = new View.OnTouchListener() {
        @Override
        public boolean onTouch(View view, MotionEvent motionEvent) {
            if (mAutoHide) {
                delayedHide(AUTO_HIDE_DELAY_MILLIS);
            }
            return false;
        }
    };

    private TextureView mSurfaceTv;
    private View mProgressView;
    private View mControllerView;
    private boolean mVisible;

    private ImageView mPlayIv;
    private ImageView mBackwardIv;
    private ImageView mForwardIv;

    private SeekBar mProgressSb;
    private TextView mCurrentProgressTv;
    private TextView mDurationTv;

    private MediaPlayer mMediaPlayer;
    private MediaPlayState mPlayState;
    private String mMediaFilePath;

    private TimeCountThread mTimeCountThread;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // 导航栏、状态栏透明
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS
                    | WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
            window.getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(getResources().getColor(R.color.colorTranslucence));
            window.setNavigationBarColor(getResources().getColor(R.color.colorTranslucence));

            if (getSupportActionBar() != null) {
                getSupportActionBar().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            }
        }

        setContentView(R.layout.activity_video_player);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        mVisible = true;


        handleIntent();
        findView();
        setView();

        initMediaPlayer();
    }

    private void handleIntent() {
        mMediaFilePath = getIntent().getStringExtra(EXTRA_MEDIA_PATH_KEY);
        if (TextUtils.isEmpty(mMediaFilePath)) {
            TipsUtil.showToast(this, "源文件位置未知");
            finish();
        }
    }

    private void findView() {
        mProgressView = findViewById(R.id.rl_progress);
        mControllerView = findViewById(R.id.rl_controller);
        mSurfaceTv = (TextureView) findViewById(R.id.tv_surface);
        mPlayIv = (ImageView) findViewById(R.id.iv_play);
        mBackwardIv = (ImageView) findViewById(R.id.iv_backward);
        mForwardIv = (ImageView) findViewById(R.id.iv_forward);
        mProgressSb = (SeekBar) findViewById(R.id.seek_bar);
        mCurrentProgressTv = (TextView) findViewById(R.id.tv_current_progress);
        mDurationTv = (TextView) findViewById(R.id.tv_duration);
    }

    private void setView() {
        mSurfaceTv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                toggle();
            }
        });
        mSurfaceTv.setSurfaceTextureListener(new TextureView.SurfaceTextureListener() {
            @Override
            public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
                Log.d(TAG, "onSurfaceTextureAvailable");
                if (mMediaPlayer == null || mPlayState == MediaPlayState.IDLE) {
                    initMediaPlayer();
                }
                mMediaPlayer.setSurface(new Surface(surface));
                playMedia();
            }

            @Override
            public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {
                Log.d(TAG, "onSurfaceTextureSizeChanged width = " + width + " height = " + height);
            }

            @Override
            public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
                Log.d(TAG, "onSurfaceTextureDestroyed");
                releaseMedia();
                return true;
            }

            @Override
            public void onSurfaceTextureUpdated(SurfaceTexture surface) {
            }
        });

        mPlayIv.setOnClickListener(this);
        mBackwardIv.setOnClickListener(this);
        mForwardIv.setOnClickListener(this);

        mProgressSb.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    mMediaPlayer.seekTo(seekBar.getProgress());

                    delayedHide(AUTO_HIDE_DELAY_MILLIS);
                }
                mCurrentProgressTv.setText(TimeUtil.formatDuration(mMediaPlayer.getCurrentPosition() / 1000));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
    }

//    @Override
//    protected void onResume() {
//        super.onResume();
//        initMediaPlayer();
//    }

    @Override
    protected void onPause() {
        super.onPause();

        if (MediaPlayState.STARTED == mPlayState) {
            pauseMedia();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        releaseMedia();
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);

        // Trigger the initial hide() shortly after the activity has been
        // created, to briefly hint to the user that UI controls
        // are available.
        delayedHide(100);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            // This ID represents the Home or Up button.
//            NavUtils.navigateUpFromSameTask(this);
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void toggle() {
        if (mVisible) {
            hide();
        } else {
            show();
        }
    }

    private void hide() {
        // Hide UI first
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.hide();
        }
        mProgressView.setVisibility(View.GONE);
        mControllerView.setVisibility(View.GONE);
        mVisible = false;

        // Schedule a runnable to remove the status and navigation bar after a delay
        mHideHandler.removeCallbacks(mShowPart2Runnable);
        mHideHandler.postDelayed(mHidePart2Runnable, UI_ANIMATION_DELAY);
    }

    @SuppressLint("InlinedApi")
    private void show() {
        // Show the system bar
        mSurfaceTv.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        mVisible = true;

        // Schedule a runnable to display UI elements after a delay
        mHideHandler.removeCallbacks(mHidePart2Runnable);
        mHideHandler.postDelayed(mShowPart2Runnable, UI_ANIMATION_DELAY);
    }

    /**
     * Schedules a call to hide() in [delay] milliseconds, canceling any
     * previously scheduled calls.
     */
    private void delayedHide(int delayMillis) {
        if (mAutoHide) {
            mHideHandler.removeCallbacks(mHideRunnable);
            mHideHandler.postDelayed(mHideRunnable, delayMillis);
        }
    }


    private void initMediaPlayer() {
        Log.d(TAG, "initMediaPlayer");
        if (mMediaPlayer == null) {
            mMediaPlayer = new MediaPlayer();
            mPlayState = MediaPlayState.IDLE;
            mMediaPlayer.setScreenOnWhilePlaying(true);
            mMediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
                @Override
                public void onCompletion(MediaPlayer mp) {
                    mPlayState = MediaPlayState.COMPLETED;

                    mPlayIv.setImageResource(R.mipmap.ic_play);
                    int padding = DensityUtil.dip2px(mContext, 15);
                    int paddingDiff = DensityUtil.dip2px(mContext, 2);
                    mPlayIv.setPadding(padding + paddingDiff, padding, padding - paddingDiff, padding);

                    mAutoHide = false;
                    mHideHandler.removeCallbacks(mHideRunnable);
                    show();
                }
            });
            mMediaPlayer.setOnVideoSizeChangedListener(new MediaPlayer.OnVideoSizeChangedListener() {
                @Override
                public void onVideoSizeChanged(MediaPlayer mp, int width, int height) {
                    Log.d(TAG, "MediaPlayer onVideoSizeChanged width = " + width + " height = " + height);

                    mCurrentProgressTv.setText(TimeUtil.formatDuration(0));
                    mDurationTv.setText(TimeUtil.formatDuration(mMediaPlayer.getDuration() / 1000));
                    mProgressSb.setProgress(0);
                    mProgressSb.setMax(mMediaPlayer.getDuration());

                    configSurfaceSize(width, height);
                }
            });
            mMediaPlayer.setOnSeekCompleteListener(new MediaPlayer.OnSeekCompleteListener() {
                @Override
                public void onSeekComplete(MediaPlayer mp) {
                    Log.d(TAG, "MediaPlayer onSeekComplete ");
                }
            });
            mMediaPlayer.setOnErrorListener(new MediaPlayer.OnErrorListener() {
                @Override
                public boolean onError(MediaPlayer mp, int what, int extra) {
                    Log.e(TAG, "MediaPlayer onError");
                    mPlayState = MediaPlayState.ERROR;

                    mp.reset();
                    mPlayState = MediaPlayState.IDLE;

                    initMediaPlayer();
                    return true;
                }
            });
        }
        setMediaPlayerSource();
        prepareMediaPlayer();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        // 这个方法很重要
        Point point = DensityUtil.getScreenRealSize(this);
        Log.d(TAG, "onConfigurationChanged size = " + point);

        if (mMediaPlayer.getVideoWidth() > 0 && mMediaPlayer.getVideoHeight() > 0) {
            configSurfaceSize(mMediaPlayer.getVideoWidth(), mMediaPlayer.getVideoHeight());
        }
    }

    private void configSurfaceSize(int width, int height) {
        ViewGroup.LayoutParams layoutParams = mSurfaceTv.getLayoutParams();
        Point point = DensityUtil.getScreenRealSize(this);
        int screenWidth = point.x;
        int screenHeight = point.y;
//        int screenWidth = mContext.getResources().getDisplayMetrics().widthPixels;
//        int screenHeight = mContext.getResources().getDisplayMetrics().heightPixels;
        if (((double) screenWidth / width) * height > screenHeight) {
            layoutParams.width = (int) (((double) screenHeight / height) * width);
            layoutParams.height = screenHeight;
        } else {
            layoutParams.width = screenWidth;
            layoutParams.height = (int) (((double) screenWidth / width) * height);
            Log.d(TAG, "onConfigurationChanged width = " + layoutParams.width + " height = " + layoutParams.height);
        }
    }

    private void setMediaPlayerSource() {
        try {
            mMediaPlayer.setDataSource(mMediaFilePath);
            mPlayState = MediaPlayState.INITIALIZED;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void prepareMediaPlayer() {
        try {
            mMediaPlayer.prepare();
            mPlayState = MediaPlayState.PREPARED;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void playMedia() {
        mPlayIv.setImageResource(R.mipmap.ic_pause);
        int padding = DensityUtil.dip2px(this, 15);
        mPlayIv.setPadding(padding, padding, padding, padding);

        mMediaPlayer.start();
        mPlayState = MediaPlayState.STARTED;

        if (mTimeCountThread == null || mTimeCountThread.mIsClosed) {
            mTimeCountThread = new TimeCountThread();
        }
        if (mTimeCountThread.mIsPaused) {
            mTimeCountThread.resumeThread();
        } else {
            mTimeCountThread.start();
        }

        mAutoHide = true;
    }

    private void pauseMedia() {
        mPlayIv.setImageResource(R.mipmap.ic_play);
        int padding = DensityUtil.dip2px(this, 15);
        int paddingDiff = DensityUtil.dip2px(this, 2);
        mPlayIv.setPadding(padding + paddingDiff, padding, padding - paddingDiff, padding);

        mMediaPlayer.pause();
        mPlayState = MediaPlayState.PAUSED;

        if (mTimeCountThread != null) {
            mTimeCountThread.pauseThread();
        }
    }

    private void stopMedia() {
        mMediaPlayer.stop();
        mPlayState = MediaPlayState.STOPPED;

    }

    private void releaseMedia() {
        if (mMediaPlayer != null) {
            if (MediaPlayState.STARTED == mPlayState || MediaPlayState.PAUSED == mPlayState) {
                stopMedia();
            }
            mMediaPlayer.release();
            mPlayState = MediaPlayState.END;
            mMediaPlayer = null;
        }

        if (mTimeCountThread != null) {
            mTimeCountThread.closeThread();
        }
    }

    @Override
    public void onClick(View v) {
        delayedHide(AUTO_HIDE_DELAY_MILLIS);
        switch (v.getId()) {
            case R.id.iv_play:
                if (MediaPlayState.PAUSED == mPlayState || MediaPlayState.COMPLETED == mPlayState) {
                    playMedia();
                } else if (MediaPlayState.STARTED == mPlayState) {
                    pauseMedia();
                }
                break;
            case R.id.iv_backward:
                if (MediaPlayState.PAUSED == mPlayState || MediaPlayState.COMPLETED == mPlayState || MediaPlayState.STARTED == mPlayState) {
                    int seekTo = mMediaPlayer.getCurrentPosition() - 10 * 1000;
                    if (seekTo > 0) {
                        mMediaPlayer.seekTo(seekTo);
                        mProgressSb.setProgress(seekTo);
                    }
                }
                break;
            case R.id.iv_forward:
                if (MediaPlayState.PAUSED == mPlayState || MediaPlayState.COMPLETED == mPlayState || MediaPlayState.STARTED == mPlayState) {
                    int seekTo = mMediaPlayer.getCurrentPosition() + 10 * 1000;
                    if (seekTo < mMediaPlayer.getDuration()) {
                        mMediaPlayer.seekTo(seekTo);
                        mProgressSb.setProgress(seekTo);
                    }
                }
                break;
        }
    }

    class TimeCountThread extends Thread {

        private boolean mIsPaused;
        private boolean mIsClosed;

        public synchronized void pauseThread() {
            mIsPaused = true;
        }

        public synchronized void resumeThread() {
            mIsPaused = false;
            this.notify();
        }

        public synchronized void closeThread() {
            try {
                notify();
                mIsClosed = true;
                interrupt();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public void run() {
            while (!mIsClosed && !isInterrupted()) {
                if (!mIsPaused) {
                    mHideHandler.post(mUpdateProgressRunnable);
                    try {
                        Thread.sleep(100);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                } else {
                    try {
                        synchronized (this) {
                            wait();
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/candy/capture/activity/VideoPlayerActivity.kt
package com.candy.capture.activity

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.graphics.Color
import android.graphics.SurfaceTexture
import android.graphics.drawable.ColorDrawable
import android.media.MediaPlayer
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.text.TextUtils
import android.view.*
import android.view.TextureView.SurfaceTextureListener
import android.view.View.OnTouchListener
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import com.candy.capture.R
import com.candy.capture.model.MediaPlayState
import com.candy.capture.util.TimeUtil
import com.candy.commonlibrary.utils.DensityUtil
import com.candy.commonlibrary.utils.LogUtil
import com.candy.commonlibrary.utils.TipsUtil
import java.io.IOException

/**
 * @Description
 * @Author zhanghaiqiang
 * @Date 2020/9/30 10:57
 */
class VideoPlayerActivity: BaseActivity(), View.OnClickListener {

    companion object {
        private const val TAG = "VideoPlayerActivity"

        const val EXTRA_MEDIA_PATH_KEY = "extra_media_path"


        /**
         * If [.mAutoHide] is set, the number of milliseconds to wait after
         * user interaction before hiding the system UI.
         */
        private const val AUTO_HIDE_DELAY_MILLIS = 3000

        /**
         * Some older devices needs a small delay between UI widget updates
         * and a change of the status and navigation bar.
         */
        private const val UI_ANIMATION_DELAY = 300
    }

    /**
     * Whether or not the system UI should be auto-hidden after
     * [.AUTO_HIDE_DELAY_MILLIS] milliseconds.
     */
    private var mAutoHide = true

    private val mHandler = Handler()
    private val mHidePart2Runnable = Runnable { // Delayed removal of status and navigation bar

        // Note that some of these constants are new as of API 16 (Jelly Bean)
        // and API 19 (KitKat). It is safe to use them, as they are inlined
        // at compile-time and do nothing on earlier devices.
        mSurfaceTv!!.systemUiVisibility = (View.SYSTEM_UI_FLAG_LOW_PROFILE
                or View.SYSTEM_UI_FLAG_FULLSCREEN
                or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)
    }
    private val mShowPart2Runnable = Runnable { // Delayed display of UI elements
        val actionBar = supportActionBar
        actionBar?.show()
        mProgressView!!.visibility = View.VISIBLE
        mControllerView!!.visibility = View.VISIBLE
        delayedHide(AUTO_HIDE_DELAY_MILLIS)
    }
    private val mHideRunnable = Runnable { hide() }
    private val mUpdateProgressRunnable: Runnable = object : Runnable {
        override fun run() {
            if (mMediaPlayer != null && isPlayable()) {
                mProgressSb!!.progress = mMediaPlayer!!.currentPosition
                mHandler.removeCallbacks(this)
                mHandler.postDelayed(this, 100)
            }
        }
    }

    /**
     * Touch listener to use for in-layout UI controls to delay hiding the
     * system UI. This is to prevent the jarring behavior of controls going away
     * while interacting with activity UI.
     */
    private val mDelayHideTouchListener = OnTouchListener { view, motionEvent ->
        if (mAutoHide) {
            delayedHide(AUTO_HIDE_DELAY_MILLIS)
        }
        false
    }

    private var mSurfaceTv: TextureView? = null
    private var mProgressView: View? = null
    private var mControllerView: View? = null
    private var mControllerVisible = false

    private var mPlayIv: ImageView? = null
    private var mBackwardIv: ImageView? = null
    private var mForwardIv: ImageView? = null

    private var mProgressSb: SeekBar? = null
    private var mCurrentProgressTv: TextView? = null
    private var mDurationTv: TextView? = null

    private var mMediaPlayer: MediaPlayer? = null
    private var mPlayState: MediaPlayState? = null
    private var mMediaFilePath: String? = null

//    private TimeCountThread mTimeCountThread;

    //    private TimeCountThread mTimeCountThread;
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // 导航栏、状态栏透明
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val window = window
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS
                    or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
            window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_LAYOUT_STABLE)
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            window.statusBarColor = resources.getColor(R.color.colorTranslucence)
            window.navigationBarColor = resources.getColor(R.color.colorTranslucence)
            if (supportActionBar != null) {
                supportActionBar!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            }
        }
        setContentView(R.layout.activity_video_player)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        mControllerVisible = true
        handleIntent()
        findView()
        setView()
        initMediaPlayer()
    }

    private fun handleIntent() {
        mMediaFilePath = intent.getStringExtra(EXTRA_MEDIA_PATH_KEY)
        if (TextUtils.isEmpty(mMediaFilePath)) {
            TipsUtil.showToast(this, "源文件位置未知")
            finish()
        }
    }

    private fun findView() {
        mProgressView = findViewById(R.id.rl_progress)
        mControllerView = findViewById(R.id.rl_controller)
        mSurfaceTv = findViewById<View>(R.id.tv_surface) as TextureView
        mPlayIv = findViewById<View>(R.id.iv_play) as ImageView
        mBackwardIv = findViewById<View>(R.id.iv_backward) as ImageView
        mForwardIv = findViewById<View>(R.id.iv_forward) as ImageView
        mProgressSb = findViewById<View>(R.id.seek_bar) as SeekBar
        mCurrentProgressTv = findViewById<View>(R.id.tv_current_progress) as TextView
        mDurationTv = findViewById<View>(R.id.tv_duration) as TextView
    }

    private fun setView() {
        mSurfaceTv!!.setOnClickListener { toggle() }
        mSurfaceTv!!.surfaceTextureListener = object : SurfaceTextureListener {
            override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
                LogUtil.d(TAG, "onSurfaceTextureAvailable")
                if (mMediaPlayer == null || mPlayState == MediaPlayState.IDLE) {
                    initMediaPlayer()
                }
                mMediaPlayer!!.setSurface(Surface(surface))
                //                playMedia();
            }

            override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
                LogUtil.d(TAG, "onSurfaceTextureSizeChanged width = $width height = $height")
            }

            override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
                LogUtil.d(TAG, "onSurfaceTextureDestroyed")
                releaseMedia()
                return true
            }

            override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {}
        }
        mPlayIv!!.setOnClickListener(this)
        mBackwardIv!!.setOnClickListener(this)
        mForwardIv!!.setOnClickListener(this)
        mProgressSb!!.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    mMediaPlayer!!.seekTo(seekBar.progress)
                    delayedHide(AUTO_HIDE_DELAY_MILLIS)
                }
                mCurrentProgressTv!!.text = TimeUtil.formatDuration(mMediaPlayer!!.currentPosition / 1000.toLong())
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
    }

//    @Override
//    protected void onResume() {
//        super.onResume();
//        initMediaPlayer();
//    }

    //    @Override
    //    protected void onResume() {
    //        super.onResume();
    //        initMediaPlayer();
    //    }
    override fun onPause() {
        super.onPause()
        if (MediaPlayState.STARTED == mPlayState) {
            pauseMedia()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        releaseMedia()
        mHandler.removeCallbacksAndMessages(null)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        // Trigger the initial hide() shortly after the activity has been
        // created, to briefly hint to the user that UI controls
        // are available.
        delayedHide(100)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.home) {
            // This ID represents the Home or Up button.
//            NavUtils.navigateUpFromSameTask(this);
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun toggle() {
        if (mControllerVisible) {
            hide()
        } else {
            show()
        }
    }

    private fun hide() {
        // Hide UI first
        val actionBar = supportActionBar
        actionBar?.hide()
        mProgressView!!.visibility = View.GONE
        mControllerView!!.visibility = View.GONE
        mControllerVisible = false

        // Schedule a runnable to remove the status and navigation bar after a delay
        mHandler.removeCallbacks(mShowPart2Runnable)
        mHandler.postDelayed(mHidePart2Runnable, UI_ANIMATION_DELAY.toLong())
    }

    @SuppressLint("InlinedApi")
    private fun show() {
        // Show the system bar
        mSurfaceTv!!.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
        mControllerVisible = true

        // Schedule a runnable to display UI elements after a delay
        mHandler.removeCallbacks(mHidePart2Runnable)
        mHandler.postDelayed(mShowPart2Runnable, UI_ANIMATION_DELAY.toLong())
    }

    /**
     * Schedules a call to hide() in [delay] milliseconds, canceling any
     * previously scheduled calls.
     */
    private fun delayedHide(delayMillis: Int) {
        if (mAutoHide) {
            mHandler.removeCallbacks(mHideRunnable)
            mHandler.postDelayed(mHideRunnable, delayMillis.toLong())
        }
    }

    private fun isPlayable(): Boolean {
        return mPlayState == MediaPlayState.PREPARED || mPlayState == MediaPlayState.STARTED || mPlayState == MediaPlayState.PAUSED
    }

    private fun initMediaPlayer() {
        LogUtil.d(TAG, "initMediaPlayer")
        if (mMediaPlayer == null) {
            mMediaPlayer = MediaPlayer()
            mPlayState = MediaPlayState.IDLE
            mMediaPlayer!!.setScreenOnWhilePlaying(true)
            mMediaPlayer!!.setOnCompletionListener {
                mPlayState = MediaPlayState.COMPLETED
                mPlayIv!!.setImageResource(R.mipmap.ic_play)
                val padding = DensityUtil.dip2px(mContext, 15f)
                val paddingDiff = DensityUtil.dip2px(mContext, 2f)
                mPlayIv!!.setPadding(padding + paddingDiff, padding, padding - paddingDiff, padding)
                mAutoHide = false
                mHandler.removeCallbacks(mHideRunnable)
                show()
            }
            mMediaPlayer!!.setOnVideoSizeChangedListener { mp, width, height ->
                LogUtil.d(TAG, "MediaPlayer onVideoSizeChanged width = $width height = $height")
                if (isPlayable()) {
                    mCurrentProgressTv!!.text = TimeUtil.formatDuration(0)
                    mDurationTv!!.text = TimeUtil.formatDuration(mMediaPlayer!!.duration / 1000.toLong())
                    mProgressSb!!.progress = mMediaPlayer!!.currentPosition
                    mProgressSb!!.max = mMediaPlayer!!.duration
                }
                configSurfaceSize(width, height)
            }
            mMediaPlayer!!.setOnSeekCompleteListener { LogUtil.d(TAG, "MediaPlayer onSeekComplete ") }
            mMediaPlayer!!.setOnErrorListener { mp, what, extra ->
                LogUtil.e(TAG, "MediaPlayer onError")
                mPlayState = MediaPlayState.ERROR
                mp.reset()
                mPlayState = MediaPlayState.IDLE
                true
            }
            mMediaPlayer!!.setOnPreparedListener {
                LogUtil.e(TAG, "MediaPlayer onPrepared")
                mPlayState = MediaPlayState.PREPARED
                playMedia()
            }
        }
        setMediaPlayerSource()
        prepareMediaPlayer()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        // 这个方法很重要
        val point = DensityUtil.getScreenRealSize(this)
        LogUtil.d(TAG, "onConfigurationChanged size = $point")
        if (mMediaPlayer!!.videoWidth > 0 && mMediaPlayer!!.videoHeight > 0) {
            configSurfaceSize(mMediaPlayer!!.videoWidth, mMediaPlayer!!.videoHeight)
        }
    }

    private fun configSurfaceSize(width: Int, height: Int) {
        val layoutParams = mSurfaceTv!!.layoutParams
        val point = DensityUtil.getScreenRealSize(this)
        val screenWidth = point.x
        val screenHeight = point.y
        if (screenWidth.toDouble() / width * height > screenHeight) {
            layoutParams.width = (screenHeight.toDouble() / height * width).toInt()
            layoutParams.height = screenHeight
        } else {
            layoutParams.width = screenWidth
            layoutParams.height = (screenWidth.toDouble() / width * height).toInt()
            LogUtil.d(TAG, "onConfigurationChanged width = " + layoutParams.width + " height = " + layoutParams.height)
        }
    }

    private fun setMediaPlayerSource() {
        try {
            mMediaPlayer!!.setDataSource(mMediaFilePath)
            mPlayState = MediaPlayState.INITIALIZED
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun prepareMediaPlayer() {
        mMediaPlayer!!.prepareAsync()
        mPlayState = MediaPlayState.PREPARING
    }

    private fun playMedia() {
        mPlayIv!!.setImageResource(R.mipmap.ic_pause)
        val padding = DensityUtil.dip2px(this, 15f)
        mPlayIv!!.setPadding(padding, padding, padding, padding)
        mMediaPlayer!!.start()
        mPlayState = MediaPlayState.STARTED
        mHandler.removeCallbacks(mUpdateProgressRunnable)
        mHandler.post(mUpdateProgressRunnable)
        mAutoHide = true
    }

    private fun pauseMedia() {
        mPlayIv!!.setImageResource(R.mipmap.ic_play)
        val padding = DensityUtil.dip2px(this, 15f)
        val paddingDiff = DensityUtil.dip2px(this, 2f)
        mPlayIv!!.setPadding(padding + paddingDiff, padding, padding - paddingDiff, padding)
        mMediaPlayer!!.pause()
        mPlayState = MediaPlayState.PAUSED
        mHandler.removeCallbacks(mUpdateProgressRunnable)
    }

    private fun stopMedia() {
        mMediaPlayer!!.stop()
        mPlayState = MediaPlayState.STOPPED
    }

    private fun releaseMedia() {
        if (mMediaPlayer != null) {
            if (MediaPlayState.STARTED == mPlayState || MediaPlayState.PAUSED == mPlayState) {
                stopMedia()
            }
            mMediaPlayer!!.release()
            mPlayState = MediaPlayState.END
            mMediaPlayer = null
        }
        mHandler.removeCallbacks(mUpdateProgressRunnable)
    }

    override fun onClick(v: View) {
        delayedHide(AUTO_HIDE_DELAY_MILLIS)
        when (v.id) {
            R.id.iv_play -> if (MediaPlayState.PAUSED == mPlayState || MediaPlayState.COMPLETED == mPlayState) {
                playMedia()
            } else if (MediaPlayState.STARTED == mPlayState) {
                pauseMedia()
            }
            R.id.iv_backward -> if (MediaPlayState.PAUSED == mPlayState || MediaPlayState.COMPLETED == mPlayState || MediaPlayState.STARTED == mPlayState) {
                val seekTo = mMediaPlayer!!.currentPosition - 10 * 1000
                if (seekTo > 0) {
                    mMediaPlayer!!.seekTo(seekTo)
                    mProgressSb!!.progress = seekTo
                }
            }
            R.id.iv_forward -> if (MediaPlayState.PAUSED == mPlayState || MediaPlayState.COMPLETED == mPlayState || MediaPlayState.STARTED == mPlayState) {
                val seekTo = mMediaPlayer!!.currentPosition + 10 * 1000
                if (seekTo < mMediaPlayer!!.duration) {
                    mMediaPlayer!!.seekTo(seekTo)
                    mProgressSb!!.progress = seekTo
                }
            }
        }
    }
}